### ISSUE TITLE
Implement a MySQL source connector

### ISSUE THREAD

**Issue by willdonnelly** [estuary/connectors#77]
I've been working on this for the past week, and while I'm not entirely done with the details the broad strokes are mostly there. Filing this bug so I can update the [Big Board](https://github.com/orgs/estuary/projects/2).

A large chunk of the logic previously written in `source-postgres` is actually the generic "SQL Database Capture" process, so the first step of this is/was to factor out all of that into a generic `sqlcapture` package, and have `source-postgres` implement the various database operations (discover tables, stream change events via replication, write watermarks, scan table chunks, etc) plus a thin `main` function that injects that concrete implementation into the generic machinery that coordinates everything.

A similar bit of refactoring allows the tests to be factored out into another package as well, because most of the existing `source-postgres` tests should be just as applicable to any other SQL database capture.

After all that, implementing the MySQL source "only" requires implementing all those concrete operations one-by-one. This is actually working at the moment, and all the "genericized" tests are passing. However I expect I'll be working on various details for at least the rest of this week.

**Comment by willdonnelly** [comment]
As of this morning the current set of unit tests all pass, although there are still several things that will need to be improved:

- The JSON schema generation logic only knows about the `int`, `varchar`, and `text` column types, because those are the only ones exercised by the generic discovery test. More unit tests should probably be added, if I can find a good list of what types are "Standard SQL". Or I can do like I did for PostgreSQL and have a database-specific test for every type the database supports.
- Similarly, the "type translation" that turns values from backfill rows and replication events into appropriate JSON output is currently very naive, and probably doesn't actually work right for types that aren't integers and text. Fixing this will likely require adding some new plumbing so that the translation can take the discovered schema into account, which wasn't necessary for PostgreSQL.
- Currently the replication side of things only works if the database is running with `--binlog-row-metadata=FULL` because it expects the `TableMap` events to include column names, and those are omitted by default. It would be nice if this requirement was removed so that the connector would work without requiring database config changes, but this requires that the table discovery information be plumbed into the replication event processing so that the list of values can be paired up with the appropriate column names. Note that this also ties into the previous point about needing information about the expected *types* for each column, as well.

Also including a random link here just so I don't lose track of it again, this documentation is the only description I've found of what the various MySQL binlog events semantics actually are: https://dev.mysql.com/doc/dev/mysql-server/latest/classbinary__log_1_1Rows__event.html

**Comment by willdonnelly** [comment]
The big problem at the moment is still the point where the replication code has to match up the list of integers/strings/nulls from a MySQL `RowsEvent` with names and datatypes, which is currently solved by running with `--binlog-row-metadata=FULL` but we'd like to not require that.

To recap, the issue is that when we get an Insert/Update/Delete event from MySQL binlog replication, it contains a list of values of different "storage classes", basically nulls, integers, floats, and bytes. We need to associate each of these values with the correct column name, and we also need to know how to format the value for JSON output (for instance, are these bytes a human-readable text string, or a byte blob that should be base64-encoded?). And the part about types is less fiddly, because if we get the field name right then it's unlikely we'll ever see its type change.

So the problem here boils down to **how can we get a list of column names, in the same order that the column values are serialized into a `RowsEvent`?**

If we could assume that table structure never changes, then we could just perform table discovery at startup (which we already do to verify state/catalog sanity) and plumb that into the replication event processing code. But in the presence of `ALTER TABLE` statements, if we're processing older binlog events, it's possible that the discovery results are substantially newer than those events.

There are three solutions to this problem:

- If `--binlog-row-metadata=FULL` then each `RowsEvent` will be preceded by a `TableMapEvent` which gives us the field names at that exact point in time. We don't want to require `--binlog-row-metadata=FULL` but it seems reasonable that if the information is available we should just use that. This could also serve as a useful escape hatch in the event that certain sorts of table alteration can't be handled by other methods.
- We could parse the query from each `QueryEvent` and transform the parse tree of each `ALTER TABLE` statement into a suitable list of column names. However this doesn't necessarily give us an initial value before we've observed any `ALTER TABLE` queries, and might also require embedding the current list of columns for each table in state outputs, which is unpleasant.
- We could go perform table discovery again whenever we see an `ALTER TABLE` statement. If we're mostly caught up to the latest replication events this should just work. But if we're processing "old events" relative to the current state of the table, then the discovery result might differ from what existed at that point in the binlog. My intuition says this is safe only if we can require or guarantee that the difference will only ever be new columns added to the end of the list.

I'm leaning towards the composite solution of recommending `--binlog-row-metadata=FULL`, falling back to table discovery metadata when that's not present, and triggering a new discovery process whenever we observe an `ALTER TABLE` statement. The only situation where this might fail is when:

1. The database is running without full metadata.
2. There's a substantial backlog of replication events, such that the connector still needs to process old changes from before the table was altered to its current form.
3. The table alteration involved removing columns, or adding columns in such a way that they aren't at the end of the list.

I still need to perform some experiments to see if/when MySQL alters the binlog serialized `RowsEvent`s in a backwards-incompatible way.

**Comment by willdonnelly** [comment]
It's worth noting in relation to point #2 above that the first time the capture is launched it begins reading the binlog from the _latest_ position according to the DB and thus gets to skip all that backlog. The only times we should ever have a meaningful replication backlog are if the connector either cannot keep up with the rate of events on the server, or if it gets killed and stays offline for a substantial length of time.

**Comment by dyaffe** [comment]
This seems like a reasonable plan to me although I admittedly don't understand the complexities around parsing the bin log -- one question I've been meaning to ask is whether this decision should be made with more context around how future connectors behave.  Ex. does how SQL Server and Oracle handle similar things affect how we think about this?  

I think the answer is probably not since the method that you're advocating can probably be applied in other connectors too.

**Comment by willdonnelly** [comment]
**TL;DR: We're going to keep the change decoding logic as naive as possible and require `binlog_row_metadata = FULL` at least for the moment.**

So after a lot more discussion and pondering, I no longer believe that "trigger discovery when an `ALTER TABLE` query is seen" is an acceptable approach. It would work in **_almost_** all circumstances, but in certain corner cases (specifically, if the connector is restarted with a backlog to work through which includes more than one incompatible alteration) it could lead to silent data corruption, which even if detected would be very painful to fix. I don't think any amount of cleverness can fix the fundamental problem there, which is that the only table schema we can learn via discovery is the "current" one.

That leaves us only two options: requiring full binlog metadata, or parsing the actual `ALTER TABLE` queries and applying updates to our own internal model of the schema. Because parsing and interpreting the `ALTER TABLE` queries is a substantial bit of work and would require a lot of verification effort to get right, I would rather avoid it until/unless it's actually required. But if/when this is actually necessary we can implement it by "just" writing a bunch of code and testing it thoroughly -- there are no fundamental issues there other than it being a lot of work.

So I believe that we should require `binlog_row_metadata = FULL` for now. Luckily there are a handful of things which make this a more reasonable approach than I had initially thought:

1. The setting can be changed live by executing the query `SET PERSIST binlog_row_metadata = 'FULL'` with appropriate permissions. No database restart or downtime is required.
2. The [relevant documentation](https://dev.mysql.com/doc/mysql-replication-excerpt/8.0/en/replication-options-binary-log.html#sysvar_binlog_row_metadata) states that this setting/metadata is intended for "External software [...] to decode row events and store the data into external databases, such as a data warehouse", so this is absolutely the intended tool for the job here.
3. In situations where a user might be unwilling to toggle that setting on their main database, we suspect it should be possible to spin up a dedicated replica with full metadata enabled, and source from that.
    - This still needs to be tested empirically, but if it works it means that the goal of "not requiring the big important database to have its settings changed" can still be accomplished even if we require full metadata for the connector.
    - It _might_ even be reasonable to build a Docker image which runs such a replica alongside the connector, but only if we can figure out a way to avoid actually storing the table contents locally.

**Comment by willdonnelly** [comment]
The initial implementation of `source-mysql` is merged, and I've filed issues for things we know will require further work:

- https://github.com/estuary/connectors/issues/92
- https://github.com/estuary/connectors/issues/93
- https://github.com/estuary/connectors/issues/94
